python/tests/catalog/test_base.py

```diff
@@ -0,0 +1,410 @@
+    # Given
+    identifier = "com.organization.department.my_table"
+    # When
+    namespace_from = Catalog.namespace_from(identifier)
+    # Then
+    assert namespace_from == ("com", "organization", "department")
+
+
+def test_name_from_tuple():
+    # Given
+    identifier = ("com", "organization", "department", "my_table")
+    # When
+    name_from = Catalog.table_name_from(identifier)
+    # Then
+    assert name_from == "my_table"
+
+
+def test_name_from_str():
+    # Given
+    identifier = "com.organization.department.my_table"
+    # When
+    name_from = Catalog.table_name_from(identifier)
+    # Then
+    assert name_from == "my_table"
+
+
+def test_create_table(catalog: InMemoryCatalog):
+    table = catalog.create_table(
+        identifier=TEST_TABLE_IDENTIFIER,
+        schema=TEST_TABLE_SCHEMA,
+        location=TEST_TABLE_LOCATION,
+        partition_spec=TEST_TABLE_PARTITION_SPEC,
+        properties=TEST_TABLE_PROPERTIES,
+    )
+    assert catalog.load_table(TEST_TABLE_IDENTIFIER) == table
+
+
+def test_create_table_raises_error_when_table_already_exists(catalog: InMemoryCatalog):
+    # Given
+    given_catalog_has_a_table(catalog)
+    # When
+    with pytest.raises(AlreadyExistsError, match="Table my_table already exists"):
+        catalog.create_table(
+            identifier=TEST_TABLE_IDENTIFIER,
+            schema=TEST_TABLE_SCHEMA,
+        )
+
+
+def test_load_table(catalog: InMemoryCatalog):
+    # Given
+    given_table = given_catalog_has_a_table(catalog)
+    # When
+    table = catalog.load_table(TEST_TABLE_IDENTIFIER)
+    # Then
+    assert table == given_table
+
+
+def test_table_raises_error_on_table_not_found(catalog: InMemoryCatalog):
+    with pytest.raises(NoSuchTableError, match="Table my_table not found"):
+        catalog.load_table(TEST_TABLE_IDENTIFIER)
+
+
+def test_drop_table(catalog: InMemoryCatalog):
+    # Given
+    given_catalog_has_a_table(catalog)
+    # When
+    catalog.drop_table(TEST_TABLE_IDENTIFIER)
+    # Then
+    with pytest.raises(NoSuchTableError, match="Table my_table not found"):
+        catalog.load_table(TEST_TABLE_IDENTIFIER)
+
+
+def test_drop_table_that_does_not_exist_raise_error(catalog: InMemoryCatalog):
+    with pytest.raises(NoSuchTableError, match="Table my_table not found"):
+        catalog.load_table(TEST_TABLE_IDENTIFIER)
+
+
+def test_purge_table(catalog: InMemoryCatalog):
+    # Given
+    given_catalog_has_a_table(catalog)
+    # When
+    catalog.purge_table(TEST_TABLE_IDENTIFIER)
+    # Then
+    with pytest.raises(NoSuchTableError, match="Table my_table not found"):
+        catalog.load_table(TEST_TABLE_IDENTIFIER)
+
+
+def test_rename_table(catalog: InMemoryCatalog):
+    # Given
+    given_table = given_catalog_has_a_table(catalog)
+
+    # When
+    new_table = "new.namespace.new_table"
+    table = catalog.rename_table(TEST_TABLE_IDENTIFIER, new_table)
+
+    # Then
+    assert table
+    assert table is not given_table
```

Comment:
Isn't this just asserting that the table isn't the same instance? Is that useful?

Reply:
@rdblue In the earlier commits I was asserting if the table name has changed, but since I had rollback any changes to the `Table` interface in this PR I resorted to checking just that the table is changing in some form post renaming. I will eventually be changing this to assert table name change once we add attributes back to `Table` in #3227

Reply:
Nevermind. I worked around it for the time being. Since the `Table` class isn't frozen yet I added an attribute `identifier` post creation in the `InMemoryCatalog` and asserting on `identifier` renaming for now. This obviously would fail, and rightly so, once we make `Table` immutable/frozen.